transiflow/interface/PETSc.py

```diff
@@ -38,61 +36,105 @@ class Interface(ParallelBaseInterface):
     The PETSc backend partitions the domain into Cartesian subdomains,
     while solving linear systems on skew Cartesian subdomains to deal
     with the C-grid discretization. The subdomains will be distributed
-    over multiple processors if MPI is used to run the application."""
+    over multiple processors if MPI is used to run the application.
+
+    PETSc parallel vectors (and matrices) are distributed such that each process owns a
+    contiguous range of indices (rows) without holes. The natural ordering is therefore
+    mapped to PETSc's ordering.
+    """
 
     def __init__(self, parameters, nx, ny, nz, dim, dof, comm=PETSc.COMM_WORLD):
         super().__init__(comm, parameters, nx, ny, nz, dim, dof)
 
         self.size_global = nx * ny * nz * dof
 
-        self.map = self.create_map()
+        self.map_natural = self.create_map()
         self.assembly_map = self.create_map(True)
-        self.ghosts = list(set(self.assembly_map.indices).difference(self.map.indices))
+
+        ghosts_natural = list(
+            set(self.assembly_map.indices).difference(self.map_natural.indices)
+        )
+
+        self.ghosts = self.ghosts_petsc(ghosts_natural)
+
+        self.index_ordering = PETSc.AO().createMapping(self.map_natural.indices)
+        self.map = self.create_map_petsc()
+
+        self.index_natural_local_permut = numpy.argsort(
+            numpy.r_[self.map_natural.indices, ghosts_natural]
+        )
+
+        self.index_ordering_assembly = PETSc.AO().createMapping(
+            numpy.r_[self.map_natural.indices, ghosts_natural].astype(PETSc.IntType),
+            numpy.r_[self.map.indices, self.ghosts].astype(PETSc.IntType),
+        )
 
         self.jac = None
 
     def vector(self):
-        vec = Vector().createGhost(self.ghosts, self.size_global, comm=self.comm)
-        vec.setLGMap(self.map)
-        return vec
+        return Vector().createGhost(self.ghosts, (len(self.map.indices), None), comm=self.comm)
 
     def vector_from_array(self, array):
-        return Vector.from_array(self.map, array, self.ghosts)
+        return Vector.from_array(self.map, array[self.map_natural.indices], ghosts=self.ghosts)
+
+    def ghosts_petsc(self, ghosts_natural):
```

Comment:
The changes in this commit should probably be split into several commits, but merged with previous commits that mention they "fix" the parallel implementation.